### PR TITLE
Fix Rust features

### PR DESCRIPTION
**Pull request by ggwpez**
### Some crates do not compile on their own

Try `cargo c -p pallet-alliance --features runtime-benchmarks`, which does not work on Substrate master.  
The fix is to always enable the std features of all non-optional dependencies that have a std feature.  
There is no regression script provided by me but it could be done. Currently I am using scripts
to compile each crate on its own to see that they work with different feature sets.  

### `runtime-benchmarks` is enabled per default

Try `cargo tree -e features | grep runtime-benchmarks` to see which crates have the feature enabled
although `--features runtime-benchmarks` was never provided.  
This comes from crates either: carelessly enabling it, or crates enabling it per default.  
Fixed by checking the deps tree ensuring that it is never enabled per default.  
I wrote a regression script which wraps that `cargo tree` command and can be in-cooperated into the
CI in a future MR.

### pallet-bags-list-fuzzer requires `runtime-benchmarks`

Fixed by creating a dedicated `fuzz` feature which gates the required functions.  

Polkadot companion: https://github.com/paritytech/polkadot/pull/5983  
Cumulus companion: https://github.com/paritytech/cumulus/pull/1607

### CODE REVIEW

**Comment by shawntabrizi** [comment]
@ggwpez you made a tool? share the link!

**Review by shawntabrizi — APPROVED** [review]

File: bin/node/runtime/Cargo.toml [hunk 1]
```diff
@@ -116,6 +116,15 @@ substrate-wasm-builder = { version = "5.0.0-dev", path = "../../../utils/wasm-bu
 default = ["std"]
 with-tracing = ["frame-executive/with-tracing"]
 std = [
+	"sp-sandbox/std",
+	"pallet-whitelist/std",
+	"pallet-session-benchmarking/std",
+	"pallet-offences-benchmarking/std",
+	"pallet-nomination-pools-benchmarking/std",
+	"pallet-election-provider-support-benchmarking/std",
```
Comment (bkchr) [inline bin/node/runtime/Cargo.toml:124]: All this benchmarking stuff is optional! 

This is wrong.

However, you can use optional features

```suggestion
	"pallet-session-benchmarking?/std",
	"pallet-offences-benchmarking?/std",
	"pallet-nomination-pools-benchmarking?/std",
	"pallet-election-provider-support-benchmarking?/std",
```

**Comment by bkchr** [comment]
> @ggwpez you made a tool? share the link!

https://github.com/hack-ink/subalfred

**Comment by gilescope** [comment]
fyi there is progress being made on cargo fmt Cargo.toml. Very much still top of my xmas list this year: https://github.com/rust-lang/rustfmt/pull/5240

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1753530

**Review by kianenigma — APPROVED** [review]
Didn't look closely, but based on the description it all looks very good. 

I have a similar request as well. If I create a new pallet now, and it has `try-runtime` feature, if I add it to a runtime and forgot to enable `try-runtime`, no normal compilation will work, but the try-runtime build will fail. Can you check for this as well? You can see a few instances of it in the diff of this PR: https://github.com/paritytech/substrate/pull/10174/files

**Review by athei — APPROVED** [review]

**Review by gilescope — APPROVED** [review]

File: bin/node/runtime/Cargo.toml [hunk 2]
```diff
@@ -116,6 +116,13 @@ substrate-wasm-builder = { version = "5.0.0-dev", path = "../../../utils/wasm-bu
 default = ["std"]
 with-tracing = ["frame-executive/with-tracing"]
 std = [
+	"sp-sandbox/std",
```
Comment (bkchr) [inline bin/node/runtime/Cargo.toml:119]: You can just remove the unused depedency, then we don't need to add it here :P

**Review by bkchr — APPROVED** [review]

File: bin/node/runtime/Cargo.toml (hunk 2 above)
Comment (ggwpez) [reply bin/node/runtime/Cargo.toml:119]: There also is [cargo udeps](https://github.com/est31/cargo-udeps) which claims to find unused crates.  
I will give it a try later.

**Review by acatangiu — APPROVED** [review]

**Pull request merged**